phpstan.neon

```diff
@@ -10,8 +11,24 @@ parameters:
         - tests
     excludePaths:
         - tests/Fixtures
+    reportMaybesInPropertyPhpDocTypes: false
     ignoreErrors:
+        - '/^Anonymous function should have native return typehint ".+"\.$/'
+        - '/^Call to static method PHPUnit\\Framework\\Assert::[a-zA-Z]+\(\) with .+ will always evaluate to true\.$/'
         - '/^Class [a-zA-Z\\]+ extends @final class PhpCsFixer\\(ConfigurationException\\InvalidConfigurationException|ConfigurationException\\InvalidFixerConfigurationException|Tokenizer\\Tokens)\.$/'
+        - '/^Construct empty\(\) is not allowed\. Use more strict comparison\.$/'
+        - '/^For loop initial assignment overwrites variable \$[a-zA-Z]+\.$/'
+        - '/^Foreach overwrites \$[a-zA-Z]+ with its key variable\.$/'
+        - '/^Foreach overwrites \$[a-zA-Z]+ with its value variable\.$/'
+        - '/^Only booleans are allowed in (&&|\|\|), .+ given on the (left|right) side\.$/'
+        - '/^Only booleans are allowed in a negated boolean, .+ given\.$/'
+        - '/^Only booleans are allowed in a ternary operator condition, .+ given\.$/'
```

Comment:
to be honest, i'm not sure if I like all those rules. doesn't make sense to me to enable the ruleset and disable bunch of rules via exceptions

Reply:
To be honest I do not understand your and SP worries - I've done similar migration about 10 times and always the plan was:

> hey, let's add PHPStan with strict rules with initial list of some exceptions and then shrink that list in smaller chunks

When I run PHPStan with those exception there is 357 errors reported - I'm not that smart to fix them right away - PR would hang for a long time and for sure have a lot of conflicts to solve.

Reply:
I'm not challenging gradual progress, i'm challenging the proposed rules in general.
Some may be good, but I have the feeling that I wouldn't blindly follow all.

eg look at next rule - `only booleans are allowed in an if condition`.
```php
function foo(): ?FooType  { ... }

$foo = foo();
if ($foo) { // <<< I honestly don't see a need to replace it with `$foo !== null`, if the type is `?FooType`
    ...
}
```

Reply:
That's a surprise, I remember your comments in reviews to update logical expression and not use `boolish` ones.

From what I like in that changes in logical expression is there are changes when the fix is really useful - if that means it would require some changes that are "meh, it not so valuable" I'm fine with that.

Reply:
maybe before raising a PR that enable some "ruleset" and them add exceptionlist to ingore bunch of the rules, we may align first which rules we want to follow.

Reply:
That's exactly how I made https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/5962 - I've updated PHPStan on `master`, copied the config from this PR, added the exceptions to minor fixes here, removed the first item from ignore list, fixed the code, then reverted PHPStan config and `dev-tools/composer.json` and raised the PR.

If we merge this PR as it is it would be much simpler and faster.

I would really like to understand what is wrong with having the exception list as this PR is already improving the code - by the code I mean what is in `src/` and `tests/` directorties, if we have the exceptions for a while let's that be.

Reply:
Currently, you set exception for "violating the rules, we **want** to slowly update the code".
From my perspective, I'm not sure if we **want** to do that update, as I'm not sure whether we should indeed follow those rules. I would love to get more alignment on it first.

Reply:
I don't see a reason not to follow all those rules. And maybe we could have the same approach for rising PHPStan analyze level to 8.

Reply:
> I'm not sure if we want to do that update, as I'm not sure whether we should indeed follow those rules

Even if it's the case I'd like then to add a comment to each exception we want to stay.